flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ExternalCatalogTable.java

```diff
@@ -0,0 +1,390 @@
+	 */
+	private final boolean isSink;
+
+	/**
+	 * Properties that describe the table and should match with a {@link TableFactory}.
+	 */
+	private final Map<String, String> properties;
+
+	public ExternalCatalogTable(
+		boolean isBatch,
+		boolean isStreaming,
+		boolean isSource,
+		boolean isSink,
+		Map<String, String> properties) {
+		this.isBatch = isBatch;
+		this.isStreaming = isStreaming;
+		this.isSource = isSource;
+		this.isSink = isSink;
+		this.properties = properties;
+	}
+
+	// ----------------------------------------------------------------------------------------------
+	// Legacy code
+	// ---------------------------------------------------------------------------------------------
+
+	/**
+	 * Reads table statistics from the descriptors properties.
+	 *
+	 * @deprecated This method exists for backwards-compatibility only.
+	 */
+	@Deprecated
+	public Optional<TableStats> getTableStats() {
+		DescriptorProperties normalizedProps = new DescriptorProperties();
+		normalizedProps.putProperties(normalizedProps);
+		Optional<Long> rowCount = normalizedProps.getOptionalLong(STATISTICS_ROW_COUNT);
+		if (rowCount.isPresent()) {
+			Map<String, ColumnStats> columnStats = readColumnStats(normalizedProps, STATISTICS_COLUMNS);
+			return Optional.of(new TableStats(rowCount.get(), columnStats));
+		} else {
+			return Optional.empty();
+		}
+	}
+
+	// ----------------------------------------------------------------------------------------------
+	// Getters
+	// ----------------------------------------------------------------------------------------------
+
+	/**
+	 * Returns whether this external table is declared as table source.
+	 */
+	public boolean isTableSource() {
+		return isSource;
+	}
+
+	/**
+	 * Returns whether this external table is declared as table sink.
+	 */
+	public boolean isTableSink() {
+		return isSink;
+	}
+
+	/**
+	 * Returns whether this external table is intended for batch environments.
+	 */
+	public boolean isBatchTable() {
+		return isBatch;
+	}
+
+	/**
+	 * Returns whether this external table is intended for stream environments.
+	 */
+	public boolean isStreamTable() {
+		return isStreaming;
+	}
+
+	// ----------------------------------------------------------------------------------------------
+
+	/**
+	 * Converts this descriptor into a set of properties.
+	 */
+	@Override
+	public Map<String, String> toProperties() {
+		return properties;
+	}
+
+	/**
+	 * Creates a builder for creating an {@link ExternalCatalogTable}.
+	 *
+	 * <p>It takes {@link Descriptor}s which allow for declaring the communication to external
+	 * systems in an implementation-agnostic way. The classpath is scanned for suitable table
+	 * factories that match the desired configuration.
+	 *
+	 * <p>Use the provided builder methods to configure the external catalog table accordingly.
+	 *
+	 * <p>The following example shows how to read from a connector using a JSON format and
+	 * declaring it as a table source:
+	 *
+	 * <p>{@code
```

Comment:
Use `<pre>{@code …}</pre>` instead. See https://reflectoring.io/howto-format-code-snippets-in-javadoc/.